test/jdk/tools/jar/JarExtractTest.java

```diff
@@ -0,0 +1,382 @@
+        // extract with -P flag without any explicit destination directory (expect the extraction to work fine)
+        final String[] args = new String[]{"-xvfP", jarPath.toString()};
+        printJarCommand(args);
+        final int exitCode = JAR_TOOL.run(System.out, System.err, args);
+        Assert.assertEquals(exitCode, 0, "Failed to extract " + jarPath);
+        final String dest = ".";
+        Assert.assertTrue(Files.isDirectory(Paths.get(dest)), dest + " is not a directory");
+        final Path d1 = Paths.get(dest, "d1");
+        Assert.assertTrue(Files.isDirectory(d1), d1 + " directory is missing or not a directory");
+        final Path d2 = Paths.get(dest, "d1", "d2");
+        Assert.assertTrue(Files.isDirectory(d2), d2 + " directory is missing or not a directory");
+        final Path f1 = Paths.get(LEADING_SLASH_PRESERVED_ENTRY);
+        Assert.assertTrue(Files.isRegularFile(f1), f1 + " is missing or not a file");
+        Assert.assertEquals(Files.readAllBytes(f1), FILE_CONTENT, "Unexpected content in file " + f1);
+        final Path f2 = Paths.get("d1/d2/../f2.txt");
+        Assert.assertTrue(Files.isRegularFile(f2), f2 + " is missing or not a file");
+        Assert.assertEquals(Files.readAllBytes(f2), FILE_CONTENT, "Unexpected content in file " + f2);
+    }
+
+    /**
+     * Tests that the {@code -P} option cannot be used during jar extraction when the {@code -C} and/or
+     * {@code --dir} option is used
+     */
+    @Test
+    public void testExtractWithDirPFlagNotAllowed() throws Exception {
+        final String expectedErrMsg = "-P option cannot be used when extracting a jar to a specific location";
+        final String tmpDir = Files.createTempDirectory(Path.of("."), "8173970-").toString();
+        final List<String[]> cmdArgs = new ArrayList<>();
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", tmpDir});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "-C", "."});
+        cmdArgs.add(new String[]{"-x", "-f", testJarPath.toString(), "-P", "--dir", "."});
+        cmdArgs.add(new String[]{"-xvfP", testJarPath.toString(), "-C", tmpDir});
+        for (final String[] args : cmdArgs) {
+            final ByteArrayOutputStream err = new ByteArrayOutputStream();
+            printJarCommand(args);
+            int exitCode = JAR_TOOL.run(System.out, new PrintStream(err), args);
+            Assert.assertNotEquals(exitCode, 0, "jar extraction was expected to fail but didn't");
+            // verify it did indeed fail due to the right reason
+            Assert.assertTrue(err.toString(StandardCharsets.UTF_8).contains(expectedErrMsg));
+        }
+    }
+
+    /**
+     * Tests that {@code jar -xvf <jarname> -C <dir>} works fine too
+     */
+    @Test
+    public void testLegacyCompatibilityMode() throws Exception {
```

Comment:
Perhaps add a DataProvider so you can test --extract as well?

Reply:
So from what I understand of the code in the jar tool the "legacy compatibility" mode stands for using the single hypen option and clubbing multiple options into one. Something like `-xvfP` instead of `-x -v -f -P` and this legacy compatibility mode isn't applicable for the long form options like `--extract`. So IMO using `--extract` here won't work. Let me know if I have misunderstood this review comment or the semantics of the legacy compatibility mode.